src/pathfinder/water_regions.cpp

```diff
@@ -0,0 +1,405 @@
+{
+	return GetWaterRegionIndex(water_region.x, water_region.y);
+}
+
+/**
+ * Returns basic water region information for the provided tile.
+ * @param tile The tile for which the information will be calculated.
+ */
+WaterRegionDesc GetWaterRegionInfo(TileIndex tile)
+{
+	return WaterRegionDesc{ GetWaterRegionX(tile), GetWaterRegionY(tile) };
+}
+
+/**
+ * Returns basic water region patch information for the provided tile.
+ * @param tile The tile for which the information will be calculated.
+ */
+WaterRegionPatchDesc GetWaterRegionPatchInfo(TileIndex tile)
+{
+	WaterRegion &region = GetUpdatedWaterRegion(tile);
+	return WaterRegionPatchDesc{ GetWaterRegionX(tile), GetWaterRegionY(tile), region.GetLabel(tile)};
+}
+
+/**
+ * Marks the water region that tile is part of as invalid.
+ * @param tile Tile within the water region that we wish to invalidate.
+ */
+void InvalidateWaterRegion(TileIndex tile)
+{
+	const int index = GetWaterRegionIndex(tile);
+	if (index > static_cast<int>(_water_regions.size())) return;
+	_water_regions[index].Invalidate();
+	for (TileIndex t : _water_regions[index]) DEBUG_addDrawRectangleInstruction(t, 1, true); // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+	DEBUG_addDrawRectangleInstruction(tile, 5, true); // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+}
+
+/**
+ * Calls the provided callback function for all water region patches
+ * accessible from one particular side of the starting patch.
+ * @param water_region_patch Water patch within the water region to start searching from
+ * @param side Side of the water region to look for neigboring patches of water
+ * @param callback The function that will be called for each neighbor that is found
+ */
+inline void VisitAdjacentWaterRegionPatchNeighbors(const WaterRegionPatchDesc& water_region_patch, DiagDirection side, TVisitWaterRegionPatchCallBack& func)
+{
+	const WaterRegion &current_region = GetUpdatedWaterRegion(water_region_patch.x, water_region_patch.y);
+
+	const TileIndexDiffC offset = TileIndexDiffCByDiagDir(side);
+	const int nx = water_region_patch.x + offset.x;
+	const int ny = water_region_patch.y + offset.y;
+
+	if (nx < 0 || ny < 0 || nx >= GetWaterRegionMapSizeX() || ny >= GetWaterRegionMapSizeY()) return;
+
+	const WaterRegion &neighboring_region = GetUpdatedWaterRegion(nx, ny);
+	const DiagDirection opposite_side = ReverseDiagDir(side);
+
+	/* Indicates via which local x or y coordinates (depends on the "side" parameter) we can cross over into the adjacent region. */
+	const TWaterRegionTraversabilityBits traversability_bits = current_region.GetEdgeTraversabilityBits(side)
+		& neighboring_region.GetEdgeTraversabilityBits(opposite_side);
+	if (traversability_bits == 0) return;
+
+	if (current_region.NumberOfPatches() == 1 && neighboring_region.NumberOfPatches() == 1) {
+		func(WaterRegionPatchDesc{ nx, ny, FIRST_REGION_LABEL }); // No further checks needed because we know there is just one patch for both adjacent regions
```

Comment:
Maybe a `return` here, and un-indenting the `else`.  Now I'm wondering whether there is something that needs to happen after the `if-else` when reading this line.